clients/drcachesim/tests/analysis_unit_tests.cpp

```diff
@@ -177,10 +183,184 @@ test_queries()
     return true;
 }
 
+bool
+test_wait_records()
+{
+#ifdef HAS_ZIP
+    std::cerr << "\n----------------\nTesting wait records\n";
+
+    static constexpr int NUM_INPUTS = 5;
+    static constexpr int NUM_OUTPUTS = 2;
+    static constexpr int NUM_INSTRS = 9;
+    static constexpr memref_tid_t TID_BASE = 100;
+    static constexpr int CPU0 = 6;
+    static constexpr int CPU1 = 9;
+    std::vector<trace_entry_t> inputs[NUM_INPUTS];
+    for (int i = 0; i < NUM_INPUTS; i++) {
+        memref_tid_t tid = TID_BASE + i;
+        inputs[i].push_back(make_thread(tid));
+        inputs[i].push_back(make_pid(1));
+        // The last input will be earlier than all others. It will execute
+        // 3 instrs on each core. This is to test the case when an output
+        // begins in the wait state.
+        for (int j = 0; j < (i == NUM_INPUTS - 1 ? 6 : NUM_INSTRS); j++)
+            inputs[i].push_back(make_instr(42 + j * 4));
+        inputs[i].push_back(make_exit(tid));
+    }
+
+    // Synthesize a cpu-schedule file with some waits in it, if run in lockstep.
+    std::string cpu_fname = "tmp_test_wait_records.zip";
+    static const char *const CORE0_SCHED_STRING = "EEE-AAA-CCCAAACCCBBB.DDD.";
+    static const char *const CORE1_SCHED_STRING = "---EEE.BBBDDDBBBDDDAAA.CCC.";
+    {
+        std::vector<schedule_entry_t> sched0;
+        sched0.emplace_back(TID_BASE + 4, 10, CPU0, 0);
```

Comment:
I forget... how does the schedule entry say that cpu0 executed only 3 instrs from E? Based on the next higher start_instrs in any other schedule_entry across all cores, which would be the one on L225, but that has start_instr = 4, so shouldn't there be 4 instrs executed from this input at cpu0?

Reply:
Instr counts are 1-based.  (Yet the 1st entry lists a 0...there was a complication with the entry listing 1 IIRC.). Added a comment.